### ISSUE TITLE
Laravel passport broken as a result of #5ce37ae2

### ISSUE THREAD

**Issue by compenginecode** [lcobucci/jwt#550]
Hello,

We're currently using Laravel Passport and it's pulled in JWT 3.4. The standard "/oauth/token" route now longer works as the newly added deprecation messages are interrupting the flow, which were added in:

https://github.com/lcobucci/jwt/commit/5ce37ae2061942019e21ad5ef3af160cb5a4ad39

Don't have any more info yet but adding just in case anyone else is experiencing this?

**Comment by agwidarsito** [comment]
It appears it's being pulled in due to the package _league/oauth2-server_

**Comment by agwidarsito** [comment]
Adding an explicit reference to an earlier version when using oauth2-server works:

`"lcobucci/jwt": "3.3.0"`

Isn't the aim of a deprecation process to not break backwards compatibility but warn the user? Seems like this is physically forcing people to adjust immediately.

**Comment by muhammad-zakir** [comment]
> Adding an explicit reference to an earlier version when using oauth2-server works:
> 
> `"lcobucci/jwt": "3.3.0"`
> 
> Isn't the aim of a deprecation process to not break backwards compatibility but warn the user? Seems like this is physically forcing people to adjust immediately.
> 
> Tracking on oauth2-server also for reference to users at: [thephpleague/oauth2-server#1156](https://github.com/thephpleague/oauth2-server/issues/1156)

this should do as a temporary solution, but the latest version that won't break because of the exact same reason for now would be `3.3.3`.

**Comment by SHELA** [comment]
The same

**Comment by hungcrush** [comment]
It took me 2 hours to fix this, thanks @muhammad-zakir Now it works well.

**Comment by lcobucci** [comment]
First of all, I'm sorry for being the source of any kind of frustration. I really hope these errors appeared in a development environment and not in a production one.

I've tried to my best efforts to document and alert everyone about the `E_USER_DEPRECATED` errors in the [changelog](https://github.com/lcobucci/jwt/releases/tag/3.4.0):

> :warning: This version also triggers E_USER_DEPRECATED errors in scenarios where we can't simply use @deprecated. Please make sure you follow the provided instructions before upgrading your production code.

> Isn't the aim of a deprecation process to not break backwards compatibility but warn the user? Seems like this is physically forcing people to adjust immediately.

The approach I took (`trigger_error()`) is the same one that PHP itself uses: trigger a deprecation message and let users configure the `error_reporting` accordingly. Here's an example of PHP behaviour: https://3v4l.org/mTAjP

If we add `trigger_error()` to that snippet, you'll see that nothing breaks and only alerts happen: https://3v4l.org/FibvK
The custom error isn't displayed, though, when `error_reporting` is adjusted: https://3v4l.org/eFajm

With that said, I don't believe this to be a breakage nor forcing people to adjust immediately.

---

### Alternative approaches

Some libraries/frameworks use other approaches to this: `@trigger_error()` or use a library that allows users to toggle the alerts.

The custom error isn't displayed when using `@trigger_error()` (https://3v4l.org/B2YmC). However, with that, the only way for people to be aware of the deprecations is by having custom error handlers (messing up with the way the language operates). I can tell many stories about this but @Ocramius is a much better storyteller than me.

Pushing an external library is something that I'm also against because it would force people to download even more things.

---

### How to "fix" the issues

We worked together with both Passport and Oauth2-server maintainers to make sure everything is properly tested and gets released (they were just waiting for me to release things).

You then have 2 options:

1. Wait for the downstream libraries to release and update everything together (there's nothing really big in v3.4 apart from the forward compatibility layer, so you don't necessarily need to update `lcobucci/jwt`)
1. Configure your `error_reporting` on production not to get undesired issues (`E_ALL & ~E_USER_DEPRECATED` should suffice to resolve this particular thing but you may want to remove other levels too)

**Comment by lcobucci** [comment]
I'll close this as `invalid` as per my comment. Please feel free to ask more questions.

**Comment by agwidarsito** [comment]
> The approach I took (trigger_error()) is the same one that PHP itself uses: trigger a deprecation message and let users configure the error_reporting accordingly. Here's an example of PHP behaviour: https://3v4l.org/mTAjP

Makes sense, sorry if my comment was snarky. Adjusting `error_reporting` is a good solution and a good way to deprecate something. I think it was just extra confusing on my end because somehow this message was bubbling quite high into the application and was hard to tell where this literal text was coming from :)

**Comment by lcobucci** [comment]
> sorry if my comment was snarky :)

It's alright, we're only humans after all. *something something **emotions** something something* 😆 

> I think it was just extra confusing on my end because somehow this message was bubbling quite high into the application and was hard to tell where this literal text was coming from :)

That's a good though, we can improve the message to add the package name or even a link to my explanation about. Would you like to send a PR to help on this? I'll be more than happy to release a v3.4.1 with improvements there.

**Comment by ngarawakimani** [comment]
Issues are good sometimes.

This issue has forced me to know that `laravel/passport` uses `thephpleague/oauth2-server` and `thephpleague/oauth2-server` uses `lcobucci/jwt`. I wouldn't have bothered to check this if everything worked fine today after I run `composer update` on my app.

**Comment by agwidarsito** [comment]
>It's alright, we're only humans after all. _something something **emotions** something something_ 😆

Haha indeed - no matter how hard we try sometime those emotions squeeze past the rational side :P

> That's a good point, though. We can improve the message to add the package name or even a link to my explanation above. Would you like to send a PR to help on this? I'll be more than happy to release a v3.4.1 with improvements there.

That sounds like a great idea - both the package name and link are good ideas, especially considering it's a deprecation notice that we expect people to resolve one way or another. I'll lodge a PR for this :)

**Comment by nickdnk** [comment]
> First of all, I'm sorry for being the source of any kind of frustration. I really hope these errors appeared in a development environment and not in a production one.
> 
> I've tried to my best efforts to document and alert everyone about the `E_USER_DEPRECATED` errors in the [changelog](https://github.com/lcobucci/jwt/releases/tag/3.4.0):
> 
> > warning This version also triggers E_USER_DEPRECATED errors in scenarios where we can't simply use @deprecated. Please make sure you follow the provided instructions before upgrading your production code.
> 
> > Isn't the aim of a deprecation process to not break backwards compatibility but warn the user? Seems like this is physically forcing people to adjust immediately.
> 
> The approach I took (`trigger_error()`) is the same one that PHP itself uses: trigger a deprecation message and let users configure the `error_reporting` accordingly. Here's an example of PHP behaviour: https://3v4l.org/mTAjP
> 
> If we add `trigger_error()` to that snippet, you'll see that nothing breaks and only alerts happen: https://3v4l.org/FibvK
> The custom error isn't displayed, though, when `error_reporting` is adjusted: https://3v4l.org/eFajm
> 
> With that said, I don't believe this to be a breakage nor forcing people to adjust immediately.
> 
> ### Alternative approaches
> Some libraries/frameworks use other approaches to this: `@trigger_error()` or use a library that allows users to toggle the alerts.
> 
> The custom error isn't displayed when using `@trigger_error()` (https://3v4l.org/B2YmC). However, with that, the only way for people to be aware of the deprecations is by having custom error handlers (messing up with the way the language operates). I can share many stories about how this goes sour but @Ocramius is a much better storyteller than me.
> 
> Pushing an external library is something that I'm also against because it would force people to download even more things.
> 
> ### How to "fix" the issues
> We worked together with both Passport and Oauth2-server maintainers to make sure everything is properly tested and gets released (they were just waiting for me to release things).
> 
> You then have 2 options:
> 
> 1. Wait for the downstream libraries to release and update everything together (there's nothing really big in v3.4 apart from the forward compatibility layer, so you don't necessarily need to update `lcobucci/jwt`)
> 2. Configure your `error_reporting` on production not to get undesired issues (`E_ALL & ~E_USER_DEPRECATED` should suffice to resolve this particular thing but you may want to remove other levels too)

Hello

Can you elaborate on "in scenarios where we can't simply use @deprecated"? Which scenarios are those?

**Comment by lcobucci** [comment]
@nickdnk sure!

When deprecating behaviour or arguments of a method. With PHP 8.0 attributes we can annotate arguments but that doesn't help older PHP versions.

Example: on v3.3.0, I've backported some methods of the new `Builder` interface. However, that was a premature move (or way too hopeful move?) and, later on, the signature of those methods have changed:

```diff
 interface Builder
 {
-     public function expiresAt(int $expiration);
+     public function expiresAt(DateTimeImmutable $expiration);
 }
```

On v3.4, that needed to be adjusted. To avoid a BC-break I've used the auto conversion + deprecation to let users know that going forward they should start using  `DateTimeImmutable` objects:

```diff
 class Builder
 {
-     /** @param int $expiration */
+     /** @param int|DateTimeImmutable $expiration */
     public function expiresAt($expiration)
     {
+        if (! $expiration instanceof DateTimeImmutable) {
+            $expiration = new DateTimeImmutable('@' . $expiration);
+
+            trigger_error('Message', E_USER_DEPRECATED);
+        }

         // do the rest
     }
 }
```

I hope this helps!

**Comment by nickdnk** [comment]
> @nickdnk sure!
> 
> When deprecating behaviour or arguments of a method. With PHP 8.0 attributes we can annotate arguments but that doesn't help older PHP versions.
> 
> Example: on v3.3.0, I've backported some methods of the new `Builder` interface. However, that was a premature move (or way too hopeful move?) and, later on, the signature of those methods have changed:
> 
> ```diff
>  interface Builder
>  {
> -     public function expiresAt(int $expiration);
> +     public function expiresAt(DateTimeImmutable $expiration);
>  }
> ```
> 
> On v3.4, that needed to be adjusted.
> To avoid a BC-break, I've used the auto conversion + deprecation to let users know that going forward they should start using `DateTimeImmutable` objects:
> 
> ```diff
>  class Builder
>  {
> -     /** @param int $expiration */
> +     /** @param int|DateTimeImmutable $expiration */
>      public function expiresAt($expiration)
>      {
> +        if (! $expiration instanceof DateTimeImmutable) {
> +            $expiration = new DateTimeImmutable('@' . $expiration);
> +
> +            trigger_error('Message', E_USER_DEPRECATED);
> +        }
> 
>          // do the rest
>      }
>  }
> ```
> 
> I hope this helps!

Okay, fair enough. But... nobody is forcing you to implement the trigger_error behavior as far as I can see, and it doesn't really help anyone: If someone is using 3.x, they're going to stay on 3.x until they *choose* to upgrade to 4.0, at which point it is their responsibility to consult the upgrade guide and make the necessary changes. In 4.0 you could just change this behavior and only leave a comment saying it "will be changed in 4.0" on the 3.0 branch. People will see that comment or they won't - it doesn't matter when it's part of the 4.0 changelog. I've never seen this done before and was very surprised why anyone would change or add code in a minor version that could cause problems, as evident from the numerous threads that seem to have started because of this. I would strongly advise against it going forward. People don't usually casually upgrade minor versions of PHP without going through some thorough testing, which is why I don't believe you can equate "this is how PHP does it" to "this is how dependency X does it" when it comes to deprecations and breaking changes. PHP has breaking changes in minor versions: (https://www.php.net/manual/en/migration74.incompatible.php as example). I would always assume that I can upgrade a minor version of a dependency and have to make zero changes regardless of my environment configuration.

I caught this in testing though, so nothing bad happened. But still.

Thanks for the library by the way. Been using it for years.

**Comment by shahlin** [comment]
I'm not using Passport but I still faced this issue in dev due to tymon's JWT package. I saw the fix over on that repo to update the jwt config in `config/jwt.php` to use the `Namshi::class` as shown here: [Fix for breaking change](https://github.com/tymondesigns/jwt-auth/issues/2059#issuecomment-733441001) 

Is this a permanent solution?

**Comment by lcobucci** [comment]
> I'm not using Passport but I still faced this issue in dev due to tymon's JWT package. I saw the fix over on that repo to update the jwt config in `config/jwt.php` to use the `Namshi::class` as shown here: [Fix for breaking change](https://github.com/tymondesigns/jwt-auth/issues/2059#issuecomment-733441001) 
> 
> Is this a permanent solution?

You're using Laravel, though, which does error handling way aggressively (as mentioned in https://github.com/tymondesigns/jwt-auth/issues/2059#issuecomment-734511158)

**Comment by tymondesigns** [comment]
> I'm not using Passport but I still faced this issue in dev due to tymon's JWT package. I saw the fix over on that repo to update the jwt config in `config/jwt.php` to use the `Namshi::class` as shown here: [Fix for breaking change](https://github.com/tymondesigns/jwt-auth/issues/2059#issuecomment-733441001)
> 
> Is this a permanent solution?

The [namshi/jose package was deprecated a while back](https://github.com/namshi/jose#deprecation-notice), hence the change in default provider to this package. So I wouldn't consider it a permanent solution. It will work as a temporary measure until I'm able to add supported alternatives and upgrade this package to v4

**Comment by shahlin** [comment]
> The [namshi/jose package was deprecated a while back](https://github.com/namshi/jose#deprecation-notice), hence the change in default provider to this package. So I wouldn't consider it a permanent solution. It will work as a temporary measure until I'm able to add supported alternatives and upgrade this package to v4

Welp, looking forward to the updates!

**Comment by lcobucci** [comment]
> But... nobody is forcing you to implement the trigger_error behavior as far as I can see, and it doesn't really help anyone

@nickdnk leaving these messages is part of https://semver.org/#how-should-i-handle-deprecating-functionality and consistent with the language behaviour.

> I would always assume that I can upgrade a minor version of a dependency and have to make zero changes regardless of my environment configuration.

I think `regardless of my environment configuration` is where our PoVs differ. An environment that has `error_reporting` set to `-1` or `E_ALL` effectively tells the language that the person who configured the system is interested in **all types** of errors (including deprecation messages).

If you aren't interested in deprecations, then you must have your system configured accordingly (`E_ALL & ~E_DEPRECATION & ~E_USER_DEPRECATION`).

Our ecosystem has great tools, like PHPUnit, that follow the PHP configuration and behaviour, even allowing you to override some of these settings for testing purposes.

That allows for great flexibility on DX:

* Dev environment can have `E_ALL` which a custom error handler that **logs** occurrences of `E_DEPRECATED` and `E_USER_DEPRECATED` so people are aware of them
    * `error_reporting=E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED` (`<ini name="error_reporting" value="8191"/>`) is set on PHPUnit config file or PHPUnit is configured not to convert deprecations to exceptions (`convertDeprecationsToExceptions="false"` on the root tag)
* Production-like environments with `error_reporting=E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED`

In summary, my whole point is: I do trust (and expect) users to configure their environments properly so they can develop and run their apps. And yes, in an environment that disregards deprecation messages, updating from 3.3.x to 3.4.0 isn't disruptive.

**Comment by nickdnk** [comment]
> > But... nobody is forcing you to implement the trigger_error behavior as far as I can see, and it doesn't really help anyone
> 
> @nickdnk leaving these messages is part of https://semver.org/#how-should-i-handle-deprecating-functionality and consistent with the language behaviour.
> 
> > I would always assume that I can upgrade a minor version of a dependency and have to make zero changes regardless of my environment configuration.
> 
> I think `regardless of my environment configuration` is where our PoVs differ. An environment that has `error_reporting` set to `-1` or `E_ALL` effectively tells the language that the person who configured the system is interested in **all types** of errors (including deprecation messages).
> 
> If you aren't interested in deprecations, then you must have your system configured accordingly (`E_ALL & ~E_DEPRECATION & ~E_USER_DEPRECATION`).
> 
> Our ecosystem has great tools, like PHPUnit, that follow the PHP configuration and behaviour, even allowing you to override some of these settings for testing purposes.
> 
> That allows for great flexibility on DX:
> 
> * Dev environment can have `E_ALL` which a custom error handler that **logs** occurrences of `E_DEPRECATED` and `E_USER_DEPRECATED` so people are aware of them
>   
>   * `error_reporting=E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED` (`<ini name="error_reporting" value="8191"/>`) is set on PHPUnit config file or PHPUnit is configured not to convert deprecations to exceptions (`convertDeprecationsToExceptions="false"` on the root tag)
> * Production-like environments with `error_reporting=E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED`
> 
> In summary, my whole point is: I do trust (and expect) users to configure their environments properly so they can develop and run their apps. And yes, in an environment that disregards deprecation messages, updating from 3.3.x to 3.4.0 isn't disruptive.

And that's your decision, of course, just know that you're solving nothing with this and creating problems for (some) people for no *useful* reason (per my previous comment with regards to major versions requiring a code review *anyway*). As I said I've never experienced this before with any other library, so take from that what you will. You may be technically correct ("the best kind" - haha), but practically I just don't see the advantage: I'd say that a comment in phpdocs informing the user that this method will only accept DateTime (or whatever it is) in the next major is adequate. I don't assume you're going to change your mind, so let's just agree to disagree about this.

I've almost completed the migration to 4.0, so we're out of the woods anyway.

**Comment by rafwell** [comment]
+1. I need downgrade to version 3.3.*, I'm using laravel 6 and php 7.2.

**Comment by Ocramius** [comment]
Locking here: already clarified in https://github.com/lcobucci/jwt/issues/550#issuecomment-734773506, so please report new issues with a test case, if required, but "me too" doesn't really help.
